Data/ByteString/Lazy/ReadWord.hs

```diff
@@ -0,0 +1,188 @@
+-- Nothing
+--
+readWord :: ByteString -> Maybe (Word, ByteString)
+readWord bs = _readWord64
+    (fromIntegral @Word64 @Word)
+    bs
+    (_qrPair @Word maxBound)
+
+-- | A variant of 'readWord' specialised to 'Word64'.
+readWord64 :: ByteString -> Maybe (Word64, ByteString)
+readWord64 bs = _readWord64
+    (id :: Word64 -> Word64)
+    bs
+    (_qrPair @Word64 maxBound)
+
+-- | A variant of 'readWord' specialised to 'Word32'.
+readWord32 :: ByteString -> Maybe (Word32, ByteString)
+readWord32 bs = _readWord64
+    (fromIntegral @Word64 @Word32)
+    bs
+    (_qrPair @Word32 maxBound)
+
+-- | A variant of 'readWord' specialised to 'Word16'.
+readWord16 :: ByteString -> Maybe (Word16, ByteString)
+readWord16 bs = _readWord64
+    (fromIntegral @Word64 @Word16)
+    bs
+    (_qrPair @Word16 maxBound)
+
+-- | A variant of 'readWord' specialised to 'Word8'.
+readWord8 :: ByteString -> Maybe (Word8, ByteString)
+readWord8 bs = _readWord64
+    (fromIntegral @Word64 @Word8)
+    bs
+    (_qrPair @Word8 maxBound)
+
+-- | Internal implementation
+_readWord64 :: (Word64 -> a)
+            -> ByteString
+            -> (Word64, Word64)
+            -> Maybe (a, ByteString)
+{-# INLINE _readWord64 #-}
+#ifdef BYTESTRING_STRICT
+_readWord64 kont !bs !qr = case digits bs qr 0 of
+    (True, !used, !acc)
+        | used > 0
+          -> let !r = kont acc
+                 !s = unrefEmpty used bs
+              in Just (r, s)
+    !_    -> Nothing
+  where
+    unrefEmpty n s =
+        let rest = BU.unsafeDrop n s
+         in if B.null rest then empty else rest
+#else
+_readWord64 kont !bs !qr = loop bs False 0
+  where
+    -- All done
+    loop s@Empty !valid !acc
+        | valid = let !r = kont acc in Just (r, s)
+        | otherwise = Nothing
+    -- skip empty segment
+    loop (Chunk (BI.BS _ 0) cs) !valid !acc
+        = loop cs valid acc
+    loop s@(Chunk c@(BI.BS _ !len) cs) !valid !acc
+        = case digits c qr acc of
+            (True, !used, !a)
+                | used == 0 -- No more digits present
+                  -> if valid
+                     then let !r = kont acc in Just (r, s)
+                     else Nothing
+                | used < len -- input not entirely digits
+                  -> let !r = kont a
+                         !c' = BU.unsafeDrop used c
+                      in Just (r, Chunk c' cs)
+                | otherwise -- read more digits from the remaining chunks
+                  -> loop cs True a
+            _     -> Nothing
+#endif
+
+-- | Process as many digits as we can, returning the additional number of
+-- digits found, the updated accumulator, and whether the input decimal did not
+-- overflow prior to processing all the provided digits (end of input or
+-- non-digit encountered).
+--
+digits :: BI.ByteString    -- ^ input chunk
+       -> (Word64, Word64) -- ^ validity bounds
+       -> Word64           -- ^ accumulator
+       -> (Bool, Int, Word64)
+{-# INLINE digits #-}
```

Comment:
It looks like after all of the inlining is said and done, this function winds up duplicated twice in every `readInt` variant (and probably once in each `readWord` variant). Given that this duplication is occurring anyway, it might be nice to use `Word` for the accumulator in the specializations for small types, for the benefit of any 32-bit builds out there.

Reply:
Are you suggesting switching a polymorphic accumulator type, that's resolved at the call site to either `Word64` in the `Word64` functions and `Word` otherwise?

I guess I can then SPECIALISE `digits` to `Word64` and `Word` and probably do fine. 

How important do you think this is?  I prefer to make the code fast on 64-bit systems, and am not too worried about optimality on 32-bit systems.